src/target/riscv/riscv-013.c

```diff
@@ -4341,77 +4326,127 @@ static int read_memory_progbuf_inner_one(struct target *target,
 /**
  * Read the requested memory, silently handling memory access errors.
  */
-static mem_access_result_t
-read_memory_progbuf(struct target *target, const riscv_mem_access_args_t args)
+static mem_access_result_t read_memory_progbuf(struct target *target,
+		const riscv_mem_access_args_t args, bool mprven)
 {
 	assert(riscv_mem_access_is_read(args));
 
-	mem_access_result_t skip_reason = mem_should_skip_progbuf(target, args);
-	if (skip_reason != MEM_ACCESS_OK)
-		return skip_reason;
-
-	LOG_TARGET_DEBUG(target, "reading %" PRIu32 " elements of %" PRIu32
-			" bytes from 0x%" TARGET_PRIxADDR, args.count, args.size, args.address);
-
-	if (dm013_select_target(target) != ERROR_OK)
-		return MEM_ACCESS_SKIPPED_TARGET_SELECT_FAILED;
-
 	select_dmi(target);
-
 	memset(args.read_buffer, 0, args.count * args.size);
 
 	if (execute_autofence(target) != ERROR_OK)
 		return MEM_ACCESS_SKIPPED_FENCE_EXEC_FAILED;
 
+	int res = (args.count == 1) ?
+			read_memory_progbuf_inner_one(target, args, mprven) :
+			read_memory_progbuf_inner(target, args, mprven);
+
+	return res == ERROR_OK ? MEM_ACCESS_OK : MEM_ACCESS_FAILED;
+}
+
+static mem_access_result_t write_memory_progbuf(struct target *target,
+		const riscv_mem_access_args_t args, bool mprven);
+
+static mem_access_result_t
+access_memory_progbuf(struct target *target, const riscv_mem_access_args_t args)
+{
+	mem_access_result_t result = mem_should_skip_progbuf(target, args);
+	if (result != MEM_ACCESS_OK)
+		return result;
+
+	const bool is_read = riscv_mem_access_is_read(args);
+	const char *const access_type = is_read ? "read" : "write";
+	LOG_TARGET_DEBUG(target, "%sing %" PRIu32 " words of %" PRIu32
+			" bytes to 0x%" TARGET_PRIxADDR, access_type, args.count,
+			args.size, args.address);
+
+	if (dm013_select_target(target) != ERROR_OK)
+		return MEM_ACCESS_SKIPPED_TARGET_SELECT_FAILED;
+
 	uint64_t mstatus = 0;
 	uint64_t mstatus_old = 0;
 	if (modify_privilege(target, &mstatus, &mstatus_old) != ERROR_OK)
 		return MEM_ACCESS_FAILED_PRIV_MOD_FAILED;
 
-	const bool mprven = riscv_virt2phys_mode_is_hw(target)
+	const bool mprven_required = riscv_virt2phys_mode_is_hw(target)
 			&& get_field(mstatus, MSTATUS_MPRV);
-	int result = (args.count == 1) ?
-		read_memory_progbuf_inner_one(target, args, mprven) :
-		read_memory_progbuf_inner(target, args, mprven);
 
-	if (mstatus != mstatus_old &&
-			register_write_direct(target, GDB_REGNO_MSTATUS, mstatus_old) != ERROR_OK)
-		return MEM_ACCESS_FAILED;
+	result = is_read ? read_memory_progbuf(target, args, mprven_required) :
+			write_memory_progbuf(target, args, mprven_required);
 
-	return (result == ERROR_OK) ? MEM_ACCESS_OK : MEM_ACCESS_FAILED;
+	if (mstatus != mstatus_old && register_write_direct(target,
+			GDB_REGNO_MSTATUS, mstatus_old) != ERROR_OK)
+		return MEM_ACCESS_SKIPPED_REG_WRITE_FAILED;
```

Comment:
There is a conflict with #1190
I believe it will be easier to merge #1190 first and then adjust this one.